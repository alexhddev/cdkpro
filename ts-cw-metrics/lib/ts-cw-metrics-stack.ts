import * as cdk from 'aws-cdk-lib';
import { Alarm, Metric } from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { LambdaSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { Construct } from 'constructs';
import { join } from 'path';

export class TsCwMetricsStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const webHookLambda = new NodejsFunction(this, 'webHookLambda', {
      runtime: Runtime.NODEJS_18_X,
      handler: 'handler',
      entry: (join(__dirname, '..', 'services', 'hook.ts'))
    });

    const alarmTopic = new Topic(this, 'TSAlarmTopic', {
      displayName: 'TsAlarmTopic',
      topicName: 'TSAlarmTopic'
    });

    alarmTopic.addSubscription(new LambdaSubscription(webHookLambda));

    const sampleAlarm = new Alarm(this, 'Ts-ApiAlarm', {
      metric: new Metric({
          metricName: 'custom-error',
          namespace: 'Custom',
          period: cdk.Duration.minutes(1),
          statistic: 'Sum',
      }),
      evaluationPeriods: 1,
      threshold: 100,
  });

  const topicAction = new SnsAction(alarmTopic);
  sampleAlarm.addAlarmAction(topicAction);
  sampleAlarm.addOkAction(topicAction);

  const apiAlarm = new Alarm(this, 'Ts-Api4xxAlarm', {
    metric: new Metric({
        metricName: '4XXError',
        namespace: 'AWS/ApiGateway',
        period: cdk.Duration.minutes(1),
        statistic: 'Sum',
        dimensionsMap: {
          "ApiName": "TS-EmplApi"
        }
    }),
    evaluationPeriods: 1,
    threshold: 1,
});

apiAlarm.addAlarmAction(topicAction);
apiAlarm.addOkAction(topicAction);

  }
}
