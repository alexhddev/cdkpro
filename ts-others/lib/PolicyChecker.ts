import { Annotations, IAspect, Stack } from "aws-cdk-lib";
import { CfnPolicy } from "aws-cdk-lib/aws-iam";
import { IConstruct } from "constructs";



export class PolicyChecker implements IAspect{

    visit(node: IConstruct): void {
        if (node instanceof CfnPolicy) {
            const resolvedDoc = Stack.of(node).resolve(node.policyDocument)
            const resolvedDocJson = JSON.stringify(resolvedDoc);

            const forbiddenAction = 'GetBucket'

            if (resolvedDocJson.search(forbiddenAction)) {
                Annotations.of(node).addWarningV2('warining',
                'Forbidden action: ' + forbiddenAction
                )
            }
        }
    }
    
}