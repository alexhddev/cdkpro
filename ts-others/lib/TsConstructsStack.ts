import { Construct } from "constructs";
import * as cdk from 'aws-cdk-lib';
import * as lambda from "aws-cdk-lib/aws-lambda";
import { S3ToLambda } from '@aws-solutions-constructs/aws-s3-lambda';

export class TsConstructsStack extends cdk.Stack {

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new S3ToLambda(this, 'test-s3-lambda', {
      lambdaFunctionProps: {
        code: lambda.Code.fromInline('console.log()'),
        runtime: lambda.Runtime.NODEJS_16_X,
        handler: 'index.handler'
      },
    });
  }
}