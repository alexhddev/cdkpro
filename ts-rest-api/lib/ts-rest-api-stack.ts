import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { join } from 'path';
import { LambdaIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { AttributeType, Billing, TableV2 } from 'aws-cdk-lib/aws-dynamodb';

export class TsRestApiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const employeesTable = new TableV2(this, 'TS-EmplTable', {
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING
      },
      billing: Billing.onDemand()     
    });

    const emplLambda = new NodejsFunction(this, 'Ts-EmplLambda', {
      runtime: Runtime.NODEJS_18_X,
      handler: 'handler',
      entry: (join(__dirname, '..', 'services', 'handler.ts')),
      environment: {
        TABLE_NAME: employeesTable.tableName
      },
    })

    employeesTable.grantReadWriteData(emplLambda);

    const api = new RestApi(this, 'TS-EmplApi');
    const emplResource = api.root.addResource('empl');

    const emplLambdaIntegration = new LambdaIntegration(emplLambda);

    emplResource.addMethod('GET', emplLambdaIntegration);
    emplResource.addMethod('POST', emplLambdaIntegration);


  }
}
