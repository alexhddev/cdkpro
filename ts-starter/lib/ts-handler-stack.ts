import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3} from 'aws-cdk-lib'
import { Runtime, Code, Function as LambdaFunction} from 'aws-cdk-lib/aws-lambda';

interface TsHandlerStackProps extends cdk.StackProps {
    coolBucket: aws_s3.Bucket
}

export class TsHandlerStack extends cdk.Stack {

    public coolBucket: aws_s3.Bucket;

    constructor(scope: Construct, id: string, props: TsHandlerStackProps) {
        super(scope, id, props);


        new LambdaFunction(this, 'CoolHandler', {
            runtime: Runtime.NODEJS_18_X,
            handler: 'index.handler',
            code: Code.fromInline(`
            exports.handler = async (event) => {
              console.log("env bucket: " + process.env.COOL_BUCKET_ARN)
            };
          `),
            environment: {
                COOL_BUCKET_ARN: props.coolBucket.bucketArn,
            },
        });



    }

}