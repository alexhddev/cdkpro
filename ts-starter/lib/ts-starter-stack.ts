import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 } from 'aws-cdk-lib'

export class TsStarterStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new aws_s3.Bucket(this, 'TsBucket', {
      lifecycleRules: [
        {
          expiration: cdk.Duration.days(3)
        }
      ]
    })

  }
}
