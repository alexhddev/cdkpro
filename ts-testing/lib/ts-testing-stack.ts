import * as cdk from 'aws-cdk-lib';
import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

export class TsSimpleStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const lambda = new cdk.aws_lambda.Function(this, 'SimpleLambda', {
      code: Code.fromInline('console.log()'),
      handler: 'index.handler',
      runtime: Runtime.NODEJS_18_X
    });

    const bucket = new cdk.aws_s3.Bucket(this, 'SimpleBucket', {
      versioned: true
    });
    bucket.grantRead(lambda)

  }
}
