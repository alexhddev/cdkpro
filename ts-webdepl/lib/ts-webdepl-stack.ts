import * as cdk from 'aws-cdk-lib';
import { Distribution, OriginAccessIdentity } from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
import { Construct } from 'constructs';
import { existsSync } from 'fs';
import { join } from 'path';

export class TsWebdeplStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const deploymentBucket = new Bucket(this, 'TsWebDeploymentBucket')

    const uiDir = join(__dirname, '..','..', 'web', 'dist');
    if (!existsSync(uiDir)) {
      console.warn('Ui dir not found: ' + uiDir);
      return;
    }

    const originIdentity = new OriginAccessIdentity(this, 'OriginAccessIdentity');
    deploymentBucket.grantRead(originIdentity);

    const distribution = new Distribution(this, 'WebDeploymentDistribution', {
      defaultRootObject: 'index.html',
      defaultBehavior: {
        origin: new S3Origin(deploymentBucket, {
          originAccessIdentity: originIdentity
        }),
      },      
    });

    new BucketDeployment(this, 'WebDeployment', {
      destinationBucket: deploymentBucket,
      sources: [Source.asset(uiDir)],
      distribution: distribution
    });

    new cdk.CfnOutput(this, 'TsAppUrl', {
      value: distribution.distributionDomainName
    })



  }
}
