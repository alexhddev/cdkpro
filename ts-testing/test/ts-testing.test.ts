import * as cdk from 'aws-cdk-lib';
import { Template, Match, Capture } from 'aws-cdk-lib/assertions';
import * as TsTesting from '../lib/ts-testing-stack';

describe('TsSimpleStack test suite', () => {

  let template: cdk.assertions.Template

  beforeAll(() => {
    const app = new cdk.App({
      outdir: 'cdk.out/test'
    });
    const stack = new TsTesting.TsSimpleStack(app, 'MyTestStack');
    template = Template.fromStack(stack);
  })

  test('Lambda runtime check', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
      Runtime: "nodejs18.x"
    });
    template.resourceCountIs('AWS::Lambda::Function', 1)
  });
  test('Lambda runtime check', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
      Runtime: Match.stringLikeRegexp('nodejs')
    });
  });

  test('Lambda bucket policy, with matchers', () => {
    template.hasResourceProperties('AWS::IAM::Policy',
      Match.objectLike({
        PolicyDocument: {
          Statement: [{
            Resource: [{
              'Fn::GetAtt': [
                Match.stringLikeRegexp('SimpleBucket'),
                'Arn'
              ]
            },
            Match.anyValue()
            ]
          }]
        }
      })
    );
  });

  test('Lambda actions, with capture', () => {
    const lambdaActionsCapture = new Capture();

    template.hasResourceProperties('AWS::IAM::Policy', {
      PolicyDocument: {
        Statement: [{
          Action: lambdaActionsCapture
        }]
      }
    })

    const expectedActions = [
      "s3:GetObject*",
      "s3:GetBucket*",
      "s3:List*"
    ]

    expect(lambdaActionsCapture.asArray()).toEqual(
      expect.arrayContaining(expectedActions)
    )

  });



})


