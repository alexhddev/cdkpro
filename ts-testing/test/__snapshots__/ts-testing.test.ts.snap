// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TsSimpleStack test suite Bucket properties with snapshot 1`] = `
{
  "SimpleBucket250AC437": {
    "DeletionPolicy": "Retain",
    "Properties": {
      "VersioningConfiguration": {
        "Status": "Enabled",
      },
    },
    "Type": "AWS::S3::Bucket",
    "UpdateReplacePolicy": "Retain",
  },
}
`;
